### PR TITLE
Uninstall the "SUSE-Manager-Proxy" product (bsc#1133215)

### PR DESCRIPTION
**Pull request by lslezak**
- Uninstall the "SUSE-Manager-Proxy" product when upgrading from SLES12 + SUMA Proxy + SUMA Branch Server
- See https://bugzilla.suse.com/show_bug.cgi?id=1133215
- Also fix displaying wrong license (SUMA Proxy 4.0 instead of SUMA Branch Server 4.0), caused by wrong sorting of the products (missing `.reverse`, so the search did not try the longest keys first)
- Some tests were not listed in `library/packages/test/Makefile.am` so they were not executed during RPM build (fortunately they were executed at least in Travis). Additionally I have sorted the file list so the diff is a bit bigger than expected.
- 4.1.69

## Related PR

- https://github.com/yast/yast-packager/pull/428
- https://github.com/yast/yast-update/pull/124
- https://github.com/yast/yast-autoinstallation/pull/505

### Correct Product License

Originally the SUSE Manager Proxy 4.0 license was displayed here.

suma_upgrade_license

### Correct Product Summary

The old SLES12-SP3 and SUSE Manager Proxy 3.2 are upgraded to the SUSE Manager Retail Branch Server 4.0 product.

suma_upgrade_summary

### Product Details

The old SLES and SUSE Manager Proxy are selected for removal, the new base product is the SUSE Manager Retail Branch Server product.

suma_upgrade_products

### CODE REVIEW

File: library/packages/src/lib/y2packager/product_upgrade.rb [hunk 1]
```diff
@@ -44,6 +44,17 @@ class ProductUpgrade
       ["openSUSE"]                                                        => "SLES"
     }.freeze
 
+    # This maps the installed products to removed products.
```
Comment (jreidinger) [inline library/packages/src/lib/y2packager/product_upgrade.rb]: maybe make it even more clear like `This maps has as keys list of installed products and as value what products will be removed. All products in key have to be installed.`

File: library/packages/src/lib/y2packager/product_upgrade.rb [hunk 2]
```diff
@@ -94,13 +105,43 @@ def new_base_product
       # @return [<String>] Product names which obsoletes this product.
       def will_be_obsoleted_by(old_product_name)
         installed = Y2Packager::Product.installed_products.map(&:name)
-        MAPPING.each_with_object([]) do |(products, obsoleted_by), a|
-          if products.include?(old_product_name) && (products - installed).empty?
-            a << obsoleted_by
+        selected_products = Y2Packager::Product.with_status(:selected).map(&:name)
+
+        old_products = MAPPING.keys.sort_by(&:size).reverse.find do |products|
```
Comment (jreidinger) [inline library/packages/src/lib/y2packager/product_upgrade.rb]: I think this tricky line deserve a comment explaining why ( e.g. why sort by size. I have to check how constant look to understand it ). and also lines below is not easy to understand as complexity is quite high.

File: library/packages/src/lib/y2packager/product_upgrade.rb [hunk 3]
```diff
@@ -94,13 +105,43 @@ def new_base_product
...
+        [MAPPING[old_products]]
+      end
+
+      # Returns the products which are upgraded by the solver but should be actually
+      # removed from the system during upgrade. It uses an internal hardcoded
+      # product mapping.
+      # @return [<String>] Product names which should be uninstalled from the system
+      def obsolete_upgrades
+        installed = Y2Packager::Product.installed_products.map(&:name)
+        UPGRADE_REMOVAL_MAPPING.each_with_object([]) do |(products, obsolete), a|
```
Comment (jreidinger) [inline library/packages/src/lib/y2packager/product_upgrade.rb]: products everywhere. What about names line `|(all_installed, to_remove), result|` ?

File: library/packages/src/lib/y2packager/product_upgrade.rb [hunk 4]
```diff
@@ -124,7 +165,7 @@ def find_by_mapping(available)
         # to find the most specific upgrade, prefer the
         # SLES + sle-module-hpc => SLE_HPC upgrade to plain SLES => SLES upgrade
         # (if that would be in the list)
-        upgrade = MAPPING.keys.sort_by(&:size).find do |keys|
+        upgrade = MAPPING.keys.sort_by(&:size).reverse.find do |keys|
```
Comment (jreidinger) [inline library/packages/src/lib/y2packager/product_upgrade.rb]: so also bug fix?

File: library/packages/test/y2packager/product_upgrade_test.rb [hunk 5]
```diff
@@ -96,4 +102,47 @@
...
+  describe ".remove_obsolete_upgrades" do
+    it "marks the obsolete products for removal" do
+      obsolete = ["SLES", "SUSE-Manager-Proxy"]
+
+      expect(described_class).to receive(:obsolete_upgrades).and_return(obsolete)
+      obsolete.each { |o| expect(Yast::Pkg).to receive(:ResolvableRemove).with(o, :product) }
+
+      described_class.remove_obsolete_upgrades
+    end
+  end
```
Comment (jreidinger) [inline library/packages/test/y2packager/product_upgrade_test.rb]: test for that bug fix of detection ( that reverse change )?

File: library/packages/src/lib/y2packager/product_upgrade.rb (hunk 4 above)
Comment (lslezak) [reply library/packages/src/lib/y2packager/product_upgrade.rb]: Yes, the very same fix.

File: library/packages/test/y2packager/product_upgrade_test.rb (hunk 5 above)
Comment (lslezak) [reply library/packages/test/y2packager/product_upgrade_test.rb]: OK, test added.

**Pull request merged**